build catalog directly from cfn templates

jaymccon opened the pull request:
## Overview

Reads templates straight from the template files in S3, removing the need to maintain `-spec.yaml` files.

Also: 

* fixes issue where plan parameter prescriptions were not being read.
* adds support for Openshift UI parameter groups

## Related Issues

Fixes #40 

## Testing

**Testing still in progress** 

* Tested provisioning rdsmysql dev plan 

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jaymccon commented:
@vsomayaji would you mind to go over this when you get a chance.

jaymccon commented on pkg/broker/constants.go:
```diff
@@ -1,18 +1,58 @@
 var CacheTTL = 1 * time.Hour
 
 var nonCfnParams = []string{
-	"aws_access_key",
```
removing tese, as the assume role provides equivalent functionality

jaymccon commented on pkg/broker/constants.go:
```diff
@@ -1,18 +1,58 @@
 	"target_account_id",
 }
 
+var nonCfnParamDefs = map[string]interface{}{
```
injecting the nonCfnParams centrally, so no need to specify them in the template.

jaymccon commented on pkg/broker/types.go:
```diff
@@ -133,4 +133,63 @@ type UpdateCataloger func(listingcache cache.Cache, catalogcache cache.Cache, bd
+	} `yaml:"Metadata,omitempty"`
+}
+
+type OpenshiftFormDefinition struct {
```
provides parameter grouping in Openshift UI, similar to cfn's `ParameterGroup`

jaymccon commented on pkg/broker/util.go:
```diff
@@ -210,10 +212,12 @@ func generateRoleArn(params map[string]string, currentAccountID string) string {
 
 	if _, ok := params["target_account_id"]; ok {
-		targetAccountID := params["target_account_id"]
+		if params["target_account_id"] != "" {
```
now that these are being set in all cases they are empty strings if not set

vsomayaji commented on pkg/broker/awsbroker.go:
```diff
@@ -120,54 +118,26 @@ func PollUpdate(interval int, l cache.Cache, c cache.Cache, bd BucketDetailsRequ
+				glog.Errorln(err)
 			} else {
-				file, err := ioutil.ReadAll(obj.Body)
+				err = templateToServiceDefinition(file, db, c, item)
```
Maybe:
```go
			file, err := getObjectBody(s3svc, bd.bucket, bd.prefix+item.Name+templatefilter)
			if err != nil {
				glog.Errorln(err)
				continue
			}
			if err := templateToServiceDefinition(file, db, c, item); err != nil {
				glog.Errorln(err)
			}
```

vsomayaji commented on pkg/broker/awsbroker.go:
```diff
@@ -241,127 +211,130 @@ func (b *AwsBroker) ValidateBrokerAPIVersion(version string) error {
+			"imageUrl":            sd.Metadata.Spec.ImageUrl,
+			"longDescription":     sd.Metadata.Spec.LongDescription,
+		},
+		PlanUpdatable: &f,
```
Instead of declaring `t` and `f`, I sometimes do:
```go
aws.Bool(false)
```

vsomayaji commented on pkg/broker/util.go:
```diff
@@ -210,10 +212,12 @@ func generateRoleArn(params map[string]string, currentAccountID string) string {
 	targetRoleName := params["target_role_name"]
 
 	if _, ok := params["target_account_id"]; ok {
```
You can replace this with:
```go
if params["target_account_id"] != "" {
```

vsomayaji commented on pkg/broker/util.go:
```diff
@@ -401,3 +405,89 @@ func getPolicyArn(outputs []*cloudformation.Output, scope string) (string, error
+			glog.Errorln(err)
+		}
+	} else {
+		//glog.Errorf("invalid service definition for %q returned", i["name"].(string))
```
Remove commented out code.

vsomayaji commented on pkg/broker/awsbroker.go:
```diff
@@ -241,127 +211,130 @@ func (b *AwsBroker) ValidateBrokerAPIVersion(version string) error {
+			for planParam, planValue := range p.ParameterValues {
+				if planParam == paramName {
+					include = false
+					prescribed[planParam] = planValue
```
Is it correct that there's no way for a plan to change the default parameter value without prescribing it? For example, if I have an RDS template with `db.t2.small` as the default `DBInstanceClass`, I can't create a `production` plan that changes that to `db.r4.large` _and_ allows it to be modified? Not a huge deal, but I'm guessing I'll see a lot more use of the `custom` plan then.

jaymccon replied:
that is correct. Currently I'm only setting things like enabling encryption (that make more sense to be enforced/prescribed than instance type) with plan overrides. Would be nice down the line to have a way to change defaults, but still allow the provisioner to choose.

jaymccon replied:
oops, I missed that we could do that already with the -spec.yaml implementation. Will add the functionality

vsomayaji commented on pkg/broker/constants.go:
```diff
@@ -1,18 +1,58 @@
+		"display_group": "AWS Account Information",
+		"title":         "AWS Region",
+		"type":          "string",
+		"default":       "us-west-2",
```
What do you think about removing the default and making `region` a required param?

jaymccon replied:
Agreed, I think I put this in to try and minimize backward incompatible changes, but considering there are other changes that break backwards incompatible changes, might as well put it in now.

jaymccon replied:
with the default removed provisioning without specifying a region param results in stack being deployed to the region specified by the `-region` cli argument. @vsomayaji What are your thoughts on this? think we should rather be failing with a missing parameter error ?

vsomayaji commented on pkg/broker/awsbroker.go:
```diff
@@ -241,127 +211,130 @@ func (b *AwsBroker) ValidateBrokerAPIVersion(version string) error {
+				openshiftFormCreate = openshiftFormAppend(openshiftFormCreate, paramName, paramValue.(map[string]interface{}))
+				createParam := paramValue.(map[string]interface{})
+				delete(createParam, "required")
+				delete(createParam, "display_group")
```
We're removing `display_group` from these parameters but not from `nonCfnParamDefs`. Should we be removing it from both?

vsomayaji commented on pkg/broker/awsbroker.go:
```diff
@@ -241,127 +211,130 @@ func (b *AwsBroker) ValidateBrokerAPIVersion(version string) error {
+					openshiftFormUpdate = openshiftFormAppend(openshiftFormUpdate, paramName, paramValue.(map[string]interface{}))
+					updateParam := paramValue.(map[string]interface{})
+					delete(updateParam, "required")
+					delete(updateParam, "display_group")
```
We may want to delete `default` from the update schema, since we don't use the template default; we default to the value used at provision-time.

jaymccon commented on pkg/broker/aws_sdk.go:
```diff
@@ -80,3 +88,20 @@ func assumeTargetRole(sess *session.Session, params map[string]string, region st
+	})
+	if err != nil {
+		return nil, err
+	} else if obj.Body == nil {
```
When testing initially I did manage to get a nil Body, but not sure what's needed to reproduce. I'm inclined to leave the check in in-place unless we can see a case where it could do harm.

jaymccon replied:
oops, I missed that we could do that already with the `-spec.yaml` implementation. Will add the functionality

jaymccon commented:
Thanks a stack for the thorough go over @vsomayaji 😄 

I think I've addressed all the issues raised. only notable change is the additional section eaf72e4c44dbf6e1c7894c9695df5a8d5b08c555 added to plan metadata and the logic to process per plan defaults 78b6f5c862d50bdffcaec95528641e4c0b975e80

vsomayaji commented on pkg/broker/awsbroker.go:
```diff
@@ -241,127 +210,142 @@ func (b *AwsBroker) ValidateBrokerAPIVersion(version string) error {
+					openshiftFormUpdate = openshiftFormAppend(openshiftFormUpdate, paramName, paramValue.(map[string]interface{}))
+					updateParam := paramValue.(map[string]interface{})
+					for _, v := range []string{"required", "display_group", "default"} {
+						delete(updateParam, v)
```
This bit's buggy, as it will delete the keys from the original `paramValue` map. Maybe something like this instead:
```go
					updateParam := make(map[string]interface{})
					for k, v := range paramValue.(map[string]interface{}) {
						if k != "required" && k != "display_group" && k != "default" {
							updateParam[k] = v
						}
					}
```

There may be a similar problem on line 261, but I'm not positive (I haven't tested the OpenShift stuff).

https://github.com/awslabs/aws-servicebroker/pull/48/files#diff-185f61f44fd42ec6610b55fc8219880cR261

jaymccon replied:
Good catch @vsomayaji. fixed as per your suggestion.  going to merge

vsomayaji commented:
@jaymccon a few comments about the `stripTemplateId` method:
1. `golint` prefers `stripTemplateID`.
2. If `err` is not `nil`, I think `reg` may be `nil` when `reg.ReplaceAllString` is called. I'd use `MustCompile` instead.
3. If you use \`\` instead of `"`, you can avoid having to escape the `\`.

```go
func stripTemplateID(description string) string {	
	re := regexp.MustCompile(`\(qs-[a-z0-9]{9}\)`)
	return strings.TrimSpace(re.ReplaceAllString(description, ""))
}
```

You could even make the pattern a global variable.
